src/main/java/org/rumbledb/runtime/control/TypeswitchRuntimeIterator.java

```diff
@@ -45,16 +45,23 @@ public TypeswitchRuntimeIterator(
     @Override
     public void open(DynamicContext context) {
         super.open(context);
+        if (this.matchingIterator != null) {
+            this.matchingIterator.close();
+        }
+        this.matchingIterator = null;
```

Comment:
I may be missing something here so please feel free to correct me if you sense a misunderstanding.

The code here looks quite suspicious to me. The iterator initialization takes place at line 52, and the class initializes the matching iterator as null at line 22. So matching iterator should be conclusively null when open() executes. To my understanding, this is not the case, this could indicate a mishandling of the state. So this should not be fixed in open() but where the mishandling occurred.

Reply:
Also a good point. Replaced with a check and exception throwing.

Reply:
Looks beautiful now. Thank you